diff --git a/lib/index.js b/lib/index.js
index 337d6e52e5fc62c26f1f46ac115505fa8e327b5c..0eb6fca3837ef3edec25dfc25f5891a0c5911cd6 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -1181,8 +1181,7 @@ function isRequest(input) {
 }
 
 function isAbortSignal(signal) {
-	const proto = signal && typeof signal === 'object' && Object.getPrototypeOf(signal);
-	return !!(proto && proto.constructor.name === 'AbortSignal');
+	return typeof signal.aborted === "boolean" && typeof signal.removeEventListener === "function";
 }
 
 /**
